### ISSUE TITLE
Extracting all iterations from model.fit object

### ISSUE THREAD

**Issue by twooldridge** [gbradburd/conStruct#37]
Hello,

Following the instructions from this closed issue (https://github.com/gbradburd/conStruct/issues/16), I was trying to access all iterations of the MCMC chain in order to look at the entire likelihood curve, rather than just the 250 samples.  However, from what I can tell, the `model.fit` object only contains those 250 samples, and not the full run information. Am I missing something? Is there a different way to access the entire run? In brief, here's my code:

```
load('out_model.fit.Robj')
df_of_draws <- as.data.frame(model.fit)
nrow(df_of_draws)
```
[1] 250

I was playing around with the rstan `extract` function and was unable to make any progress, so any suggestions would be very much appreciated. 

Thanks in advance for your help!

**Comment by gbradburd** [comment]
The `conStruct` function calls `rstan::sampling` with the argument `save_warmup` set to `FALSE`.  So, the warmup iterations aren't actually included in the `model.fit`object.  If you want to save the whole output, you could try redefining the `conStruct` function after you've loaded the conStruct library, setting `save_warmup` to `TRUE` (see below). I'm not sure if that would break internal conStruct functions downstream.  If this does work, I'd encourage you to interpret the warmup iterations with caution.  When I was developing conStruct, I often found super wonky behavior in the first half of a run, so I followed rstan's recommendations and didn't save them.  If you're worried about assessing convergence on the stationary distribution, you're probably better off running multiple independent runs (discarding warmup for each) rather than collecting more iterations for each run.

hope that helps and let me know if you have further questions!

```
## load construct
library(conStruct)

## redefine conStruct function
conStruct <- function (spatial = TRUE, K, freqs, geoDist = NULL, coords, prefix = "", n.chains = 1, n.iter = 1000, make.figs = TRUE, save.files = TRUE,...) 
{
    call.check <- conStruct:::check.call(args <- as.list(environment()))
    freq.data <- conStruct:::process.freq.data(freqs)
    data.block <- conStruct:::make.data.block(K, freq.data, coords, spatial, 
        geoDist)
    if (save.files) {
        save(data.block, file = paste0(prefix, "_data.block.Robj"))
    }
    stan.model <- conStruct:::pick.stan.model(spatial, K)
    model.fit <- rstan::sampling(object = stanmodels[[stan.model]], 
        refresh = min(n.iter/10, 500), data = data.block, iter = n.iter, 
        chains = n.chains, thin = ifelse(n.iter/500 > 1, n.iter/500, 
            1), save_warmup = FALSE, ...)
    conStruct.results <- conStruct:::get.conStruct.results(data.block, model.fit, 
        n.chains)
    data.block <- conStruct:::unstandardize.distances(data.block)
    if (save.files) {
        save(data.block, file = paste0(prefix, "_data.block.Robj"))
        save(model.fit, file = paste(prefix, "model.fit.Robj", 
            sep = "_"))
        save(conStruct.results, file = paste(prefix, "conStruct.results.Robj", 
            sep = "_"))
    }
    if (make.figs) {
        conStruct:::make.all.the.plots(conStruct.results, data.block, prefix, 
            layer.colors = NULL)
    }
    return(conStruct.results)
}
```

**Comment by twooldridge** [comment]
Hi Gideon,

Thank you for the quick reply and the modified code! I see your point, I think I'm just used to assessing the whole run (as if I were tree-building in BEAST or something) which is why I wanted to see the entire trace. 

As far as multiple independent runs, my approach has been to submit several individuals jobs for each K, and set `n.chain=1`, rather than something like `n.chain=4`. My thinking is that I can save time by parallelizing this way (no need to run several chains consecutively), and I still get multiple independent distributions. Does this approach seem sound to you? If so, what would you recommend in terms of analyzing and combining the output? I can eyeball independent runs with `compare.two.runs` and see that structure plots, pie maps, and traces look roughly similar, but how similar is similar enough to confirm convergence? Is there an appropriate way to merge distributions across runs to get a sort of consensus distribution?

I realize this is getting into territory that is different from the original question, so I can post this as a separate issue if needed. Thanks again for your help!

Best,
Brock

**Comment by gbradburd** [comment]
Yeah running in parallel is reasonable and should save you time if you have access to a cluster.  Just make sure you're setting different seeds so the runs aren't identical!

To analyze the output, I think eye-balling is reasonable (although note that label-switching will make that tricky for layer-specific and admixture parameters if you don't use the `compare.two.runs` function).  There are more formal heuristics for assessing MCMC convergence (e.g., the Gelman-Rubin diagnostic, which compares the variance of a parameter in each independent chain to the variance of the parameter in the concatenated chains), but I haven't implemented them, so you'd have to roll your own (same with a function for merging output across runs).

**Comment by twooldridge** [comment]
Got it, I think for now I'll stick to just eyeballing the runs/using `compare.two.runs` and looking for convergence, rather than doing anything more formal. Thank you for your help!

**Comment by gbradburd** [comment]
Ok great - am I good to close this issue?

**Comment by twooldridge** [comment]
Yup, thanks again!
